apps/accounts/helpers.py

```diff
@@ -1,13 +1,48 @@
+import base64
+import json
+import os
+
 from django.conf import settings
 
+import jinja2
 from jingo import register
 
 
 @register.function
-def fxa_config():
-    return {camel_case(key): value
-            for key, value in settings.FXA_CONFIG.iteritems()
-            if key != 'client_secret'}
+@jinja2.contextfunction
+def fxa_config(context):
+    request = context['request']
+    config = {camel_case(key): value
+              for key, value in settings.FXA_CONFIG.iteritems()
+              if key != 'client_secret'}
+    set_fxa_state(request)
+    config['state'] = request.session['fxa_state']
+    return config
+
+
+def set_fxa_state(request):
+    if needs_fxa_state(request):
+        request.session['fxa_state'] = base64.urlsafe_b64encode(
+            json.dumps(fxa_state(request.user)))
+
+
+def needs_fxa_state(request):
+    def load_fxa_state():
+        return json.loads(
+            base64.urlsafe_b64decode(str(request.session['fxa_state'])))
+
+    try:
+        return ('fxa_state' not in request.session or
+                load_fxa_state()['user_id'] != request.user.pk)
+    except ValueError:
+        return True
+
+
+def fxa_state(user):
+    return {
+        'user_id': user.pk,
```

Comment:
From above: I had only included the user ID so that I knew when to invalidate it. That should likely be handled by django already though so maybe I can pull it out.